sycl/doc/Assert.md

```diff
@@ -278,149 +350,72 @@ states of definedness of `NDEBUG` macro available:
-code requires for `isAssertEnabled` property set being present in device image
-descriptor structure.
+fail.
 
+### Raising assert failure flag and reading it on host
 
-### Storing accessor metadata and writing assert failure to buffer
+Each and every translation unit provided by user should have declaration of
+assert flag read function available:
+```c++
+int __devicelib_assert_read(void);
+```
+Also, the [AssertHappened](#prog-scope-var-decl) structure type should be
+available for the copier kernel.
 
-Both storing of accessor metadata and writing assert failure is performed with
-help of built-ins. Implementations of these builtins are substituted by
-frontend.
+The definition is only provided within devicelib along with
+`__devicelib_assert_fail` function which raises the flag.
 
-User's kernel is executed through a wrapper. Wrapping takes place in DPCPP
-Runtime headers in a following manner:
+Reading of assert failure flag is performed with the help of auxiliary kernel
+which is enqueued as dependent on user's one. The flag state is checked later
+in host-task. This is achieved with approximately the following changes:
 
-```
-class handler {
+```c++
+#include <assert_happened.hpp> // contains extern decl of AssertHappenedMem
 
-template <typename KernelName> parallel_for(KernelFunc, Range) {
 #ifndef NDEBUG
-  // Assert required
-  if (!MQueue->get_device()->assert_fail_supported()) {
-    using KName2 = class ASSERT_WRAPPER_NAME(KernelName);
-    
-    auto AssertBufferAcc = MQueue->get_context()->getAssertBufferAccessor(this);
-
-    parallel_for_impl<KName2>(
-      Range,
-      [=](Item) {
-        __store_acc(AssertBuffAcc);
-        KernelFunc(Item);
-      });
-  } else {
+class AssertFlagCopier;
 #endif
 
-     // (No assert required) OR (Assert supported by device)
-     // ordinary enqueue process
-
+class queue {
+  template <typename T> event submit(T CGF) {
+    event Event = submit_impl(CGF);
 #ifndef NDEBUG
-  }
+    std::string KernelName = /* get kernel name from calls to parallel_for, etc. */;
+    // assert required
+    if (!get_device()->assert_fail_supported() && isAssertUsed(KernelName)) {
+      // __devicelib_assert_fail isn't supported by Device-side Runtime
+      // Linking against fallback impl of __devicelib_assert_fail is performed
+      // by program manager class
+      AssertHappened *AH = new AssertHappened;
+      buffer<AssertHappened, 1> *Buffer = new buffer<AssertHappened, 1>{1, AH};
+
+      // read flag value
+      event CopierEv = submit_impl([&](handler &CGH) {
+        CGH.depends_on(Event);
+
+        auto Acc = Buffer->get_access<access::mode::write>(CGH);
+
+        CGH.single_task<AssertFlagCopier>([=] {
+          Acc[0].Flag = atomic_load(&AssertHappenedMem.Flag);
+        });
+      });
+
+      // check flag state
+      submit_impl([=](handler &CGH) {
+        CGH.depends_on(CopierEv);
+
+        CGH.codeplay_host_task([=] {
+          if (AH->Flag)
```

Comment:
Should we review code snippets or they just show high level idea?

Reply:
The code snippets only show general idea

Reply:
Anyway, I'll fix obvious syntax issues.